eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import numpy as np
+import keras
+import keras.backend as K
+from keras.models import Model
+from keras.layers import Layer, Lambda
+from keras.preprocessing.image import load_img, img_to_array, array_to_img
+
+from eli5.base import Explanation
+from eli5.explain import explain_prediction
+
+
+DESCRIPTION_KERAS = """Grad-CAM visualization for image input; output is images"""
+
+# note that keras.models.Sequential subclasses keras.models.Model, so we can just register Model
+@explain_prediction.register(Model)
+def explain_prediction_keras(estimator, doc,
+                             target_names=None, # TODO: implement this
+                             targets=None,
+                             # new parameters:
+                             layer=None,
+                            ):
+    """
+    Explain an image prediction of a Keras image classifier.
+
+    See :func:`eli5.explain_prediction` for more information about the ``estimator``,
+    ``doc``, ``target_names``, and ``targets`` parameters.
+
+    Parameters
+    ----------
+    estimator : object
+        Instance of a Keras neural network model.
+    doc : object
+        An input image as a tensor to ``estimator``, for example ``numpy.ndarray``.
+    target_names : list, optional
+        *`Not Implemented`*.
+
+        Names for classes in the final output layer.
+    targets : list[int], optional
+        Prediction ID's to focus on.
+
+        *`Currently only the first prediction from the list is explained`*.
+
+        If None, the model is fed the input and its top prediction 
+        is taken as target as the target automatically.
+    layer : int or str or object, optional
+        The activation layer in the model to perform Grad-CAM on,
+        a valid keras layer name, layer index, or an instance of keras.layers.Layer.
+        
+        If None, a suitable layer is attempted to be retrieved (raise ValueError if can not).
+
+    Returns
+    -------
+    A :class:`base.Explanation` object with the ``image`` and ``heatmap`` attributes set.
+    """
+    activation_layer = get_activation_layer(estimator, layer)
+    predicted = get_target_prediction(estimator, doc, targets)
+    
+    heatmap = grad_cam(estimator, doc, predicted, activation_layer)
+    # TODO: consider renaming 'heatmap' to 'visualization'/'activations' (the output is not yet a heat map)
+
+    # need to insert a 'channel' axis to get a rank 3 image from rank 2
+    heatmap = np.expand_dims(heatmap, axis=-1)
+    heatmap = array_to_img(heatmap) # -> PIL image
+    
+    # take the single image from the input 'batch' (rank 4 to rank 3)
+    doc = doc[0]
+    image = array_to_img(doc) # -> PIL image
```

Comment:
I think it's fine as-is, but in the future we may want to support passing image and neural network input separately - e.g. if the network has some custom input (e.g. 4 channel image, 1 channel image, or something completely custom), then it might be more convenient for the user to pass both input, and also the image which corresponds to this input, which they'll prepare themselves.

Reply:
Just to clarify this idea a bit more. Correct me if I'm wrong please. We want `explain_prediction()` to be able to take two arguments:
1. One is an input to the network, for example a numpy array / tensor. We feed this input directly into the network and do calculations with it.
2. The second argument is a "displayable" image that corresponds to the input argument, for example a PIL Image object. We don't do anything with this image in `explain_prediction()`, but we can use it in `format_as_image()` to overlay the heatmap over it?

This fixes the problem when the actual input tensor is complex and can not be simply converted to a PIL image by the above two lines of code.

Reply:
@teabolt yes, this is exactly what I had in mind 👍

Reply:
@teabolt yes, this is exactly what I had in mind 👍

Reply:
@lopuhin it might be a good idea to add this straight away to not break backwards compatibility (`image` will become a "required" argument for image explanations?)

Reply:
@teabolt I assumed we could still allow passing ``doc`` only, or do you want to make it required? Either way I don't think it's a blocker for merging this PR.

Reply:
I think there's a couple of advantages for making `image` required:
* We don't need code to convert `doc` to an image, which could fail and therefore needs error handling.
* If `image` is passed, we can straight away dispatch to the image explanation function. Without `image` we would have to use the rank of `doc` or something else to determine what exactly are we explaining (image or text).

However, if you think this PR is fine without the extra argument, go ahead. We can add this later.